smartsim/_core/utils/security.py

```diff
@@ -0,0 +1,286 @@
+    """The public key"""
+
+    private: bytes = dataclasses.field(default=b"", repr=False)
+    """The private key"""
+
+    @property
+    def empty(self) -> bool:
+        """Return `True` if the KeyPair has no key values set. Useful
+        for faking the null object pattern"""
+        return self.public == self.private and len(self.public) == 0
+
+
+class KeyLocator:
+    """Determines the paths to use when persisting a `KeyPair` to disk"""
+
+    def __init__(
+        self,
+        root_dir: pathlib.Path,
+        filename: str,
+        category: str,
+        separate_keys: bool = True,
+    ) -> None:
+        """Initiailize a `KeyLocator`
+
+        :param root_dir: root path where keys are persisted to disk
+        :type root_dir: pathlib.Path
+        :param filename: the stem name of the key file
+        :type filename: str
+        :param category: the category or use-case for the key (e.g. server)
+        :type category: str
+        :param separate_keys: flag indicating if public and private keys should
+        be persisted in separate, corresponding directories
+        :type separate_keys: bool
+
+        """
+        # constants for standardized paths.
+        self._public_subdir = "pub"
+        """The category subdirectory to use when persisting a public key"""
+
+        self._private_subdir = "priv"
+        """The category subdirectory to use when persisting a private key"""
+
+        self._public_extension = "key"
+        """The extension found on public keys"""
+
+        self._private_extension = "key_secret"
+        """The extension found on private keys"""
+
+        self._key_root_dir = root_dir
+        """Path to the root directory containing key files"""
+
+        self._filename = filename
+        """Base name for key files"""
+
+        self._separate_keys = separate_keys
+        """Flag indicating if public and private keys are persisted separately"""
+
+        self._category = category
+        """(optional) Category name used to further separate key locations"""
+
+    @property
+    def public_dir(self) -> pathlib.Path:
+        """Target directory for the public key"""
+        return self.public.parent
+
+    @property
+    def private_dir(self) -> pathlib.Path:
+        """Target directory for the private key"""
+        return self.private.parent
+
+    @property
+    def public_filename(self) -> str:
+        """Filename (<stem>.<suffix>) of the public key file"""
+        return f"{self._filename}.{self._public_extension}"
+
+    @property
+    def private_filename(self) -> str:
+        """Filename (<stem>.<suffix>) of the private key file"""
+        return f"{self._filename}.{self._private_extension}"
+
+    @property
+    def public(self) -> pathlib.Path:
+        """Full target path of the public key file"""
+        # combine the root and key type (e.g. /foo/bar + /server)
+        path = self._key_root_dir / self._category  # self._filename
```

Comment:
Looks like this is left over. Is it still needed? If not feel free to remove!
```suggestion
        path = self._key_root_dir / self._category
```

Reply:
Done!